### PR TITLE
Add Fill-in-the-middle training objective example - PyTorch

### PR DESCRIPTION
**Pull request by tanaymeh**
# What does this PR do?
This PR adds a Fill-in-the-middle training objective example to 🤗 transformers.

FIM objective was proposed in [Efficient Training of Language Models to Fill in the Middle](https://arxiv.org/abs/2207.14255). They showed that autoregressive language models can learn to infill text after applying a straightforward transformation to the dataset, which simply moves a span of text from the middle of a document to its end.

As discussed in #27059 

## Who can review?
@sayakpaul @ArthurZucker

### CODE REVIEW

**Comment by tanaymeh** [comment]
This PR is currently in draft mode until I can get a GPU VM (shouldn't take more than a week at max) to test to if a model is training all fine using this script.

**Comment by tanaymeh** [comment]
@sayakpaul @ArthurZucker This PR is ready for review. I have added two scripts: `run_fim.py` and `run_fim_no_trainer.py`. Both have almost the same code structure as `run_clm.py` and `run_clm_no_trainer.py` respectively.

I have adapted the FIM dataset transformation from the [SantaCoder repository](https://github.com/loubnabnl/santacoder-finetuning/blob/main/fim.py#L22C13-L83) with changes to make it work with dataset's `.map()` function.

Training in both scripts works but I have yet to train a proper model on a GPU to compare the performance. Let me know if I need to make any changes to the script.

**Comment by tanaymeh** [comment]
@ArthurZucker @pacman100 Are there any issues in this PR? Kindly let me know!

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:130]: let's not pass this as it's deprecated
```suggestion
    token: str = field(default=None,metadata={"help": "The `token` blah blah"},
```

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:264]: this can just be the pad token.

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:512]: we should support `attn_implementation` and wait for it to be merged. 
I think we should try to incorporate as many changes as possible from latest release. Like gradient checkpointing argument, the pad to multiple of when resizing the token embedding to optimize for the architecture

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:520]: resizing should be handled like this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:527]: duplicated?

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:726]: trainer or sft trainers could be a nice to support as well

**Review by ArthurZucker — COMMENTED** [review]
thanks for this! Could you also add maybe and example log from the run to the readme>?

File: examples/pytorch/language-modeling/README.md
Comment (tanaymeh) [reply examples/pytorch/language-modeling/README.md]: I have removed this in the latest commit!

**Comment by tanaymeh** [comment]
> thanks for this! Could you also add maybe and example log from the run to the readme>?

@ArthurZucker, Since the training log is almost entirely the same as `run_clm.py` (the only difference is the dataset transformation for adding FIM tokens), should I instead add the parts of the log concerned with the training part only?

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [reply examples/pytorch/language-modeling/run_fim.py:506]: Next week 😉 q

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [reply examples/pytorch/language-modeling/run_fim.py:514]: No what I mean is using the average of the already available tokens not necessarly wte. You can always get the embeding with `get_input_embeddings` and `set_input_embeddings`

**Comment by ArthurZucker** [comment]
Sure 🤗

File: examples/pytorch/language-modeling/run_fim.py
Comment (tanaymeh) [reply examples/pytorch/language-modeling/run_fim.py:514]: @ArthurZucker I am doing this for adding new tokens' embeddings: they are set to the average of the existing embeddings. Is this a good approach?

```python
model.resize_token_embeddings(len(model.get_input_embeddings().weight) + len(new_tokens))

# Get the updated embedding layer
embeddings = model.get_input_embeddings()

# Calculate the average of the existing embeddings
avg_embedding = embeddings.weight[:-len(new_tokens)].mean(dim=0)
new_embeddings = embeddings.weight.clone()

# Set the new tokens' embeddings to this average
new_embeddings[-len(new_tokens):] = avg_embedding

# Update the model's embedding layer
embeddings.weight = torch.nn.Parameter(new_embeddings)
model.set_input_embeddings(embeddings)
```

Comment (ArthurZucker) [reply examples/pytorch/language-modeling/run_fim.py:531]: It is, but this will make the embeddings equal, you need to sample from a uniform distribution centered around the mean of the embeddings! But all good otherwise

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [reply examples/pytorch/language-modeling/run_fim.py:523]: Release done!

**Comment by tanaymeh** [comment]
@ArthurZucker I added `MultivariateNormal` distribution-based sampling of new token embeddings as was in the blog.

I have a question, should this normalisation of new embeddings also include the case when we are using `pad_to_mulitple_of`?

For example, if `pad_to_multiple_of` increases the embedding size by 10, should 10 be now added to the existing new 4 tokens we are adding, requiring us to sample 14 new token embeddings and add them to the embedding layer?

File: examples/pytorch/language-modeling/run_fim.py
Comment (tanaymeh) [reply examples/pytorch/language-modeling/run_fim.py:531]: Thanks, @ArthurZucker , I have a doubt,
When using `pad_to_mulitple_of` argument, should the tokens `len(new_tokens)` actually reflect all the new tokens added, including the ones added due to padding?

For example, if `4` FIM tokens and extra `3` tokens were added to GPT2 to make the embedding size `50264` (multiple of 8), should the embeddings' sampling be done all 7 tokens or just the original `4` FIM tokens only?

Comment (ArthurZucker) [reply examples/pytorch/language-modeling/run_fim.py:531]: I think the documentation should explain this but the idea is you just say you want len(tokenizer), the pad_to_multiple of has nothing to do with this. You should set it according to the device you are using

Comment (tanaymeh) [reply examples/pytorch/language-modeling/run_fim.py:531]: I see, thanks for clarifying this!

**Comment by tanaymeh** [comment]
@ArthurZucker I have added the `pad_to_multiple_of` argument based on the device (for GPU it will be 8 and for TPUs, it will be 128), you can see it [here](https://github.com/tanaymeh/transformers/blob/98a5976cc555696c2ae0e4eb891b9e77d989c291/examples/pytorch/language-modeling/run_fim.py#L521-L533).

I have added `attn_implementation` support in the models ([here](https://github.com/tanaymeh/transformers/blob/98a5976cc555696c2ae0e4eb891b9e77d989c291/examples/pytorch/language-modeling/run_fim.py#L504)) and the model's embedding resizing changes as you requested ([here](https://github.com/tanaymeh/transformers/blob/98a5976cc555696c2ae0e4eb891b9e77d989c291/examples/pytorch/language-modeling/run_fim.py#L529-L560)).

Please let me know if these are the correct changes or do I need to make any amends!

**Comment by ArthurZucker** [comment]
Hey! Sorry for the delay, it's a big PR I'll review it later this week 🤗

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:3]:
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:23]:
```suggestion
# You should adapt this script on your own causal language modeling task. Pointers for this are left as comments.
```

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:155]: nice

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:165]:
```suggestion
        default="sdpa", metadata={"help": ("The attention implementation to use. ")}
```
to make sure it's explicit

File: examples/pytorch/language-modeling/run_fim.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim.py:537]: nice 🤗 !

File: examples/pytorch/language-modeling/run_fim_no_trainer.py
Comment (ArthurZucker) [inline examples/pytorch/language-modeling/run_fim_no_trainer.py:3]:
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the great work here! Comments are good and should help. he way new tokens are dealt is also very nice. LGTM! 🤗

**Review by ArthurZucker — APPROVED** [review]
Let's fix the few nits and good to merge

**Comment by tanaymeh** [comment]
@ArthurZucker Thanks for your review! I have added your suggested changes and also made sure all CIs are now green.

Just waiting for @pacman100's review to get this merged!

File: examples/pytorch/language-modeling/run_fim.py
Comment (pacman100) [inline examples/pytorch/language-modeling/run_fim.py:684]:
```suggestion
        # If your application requires custom attention mask, please adjust this function's below line.
        # Since FIM transformation increases the number of tokens in input_ids and labels
```

**Review by pacman100 — COMMENTED** [review]
Thank you @tanaymeh for the example scripts on FIM Task, overall LGTM! 🔥🚀✨

Just that the embedding resizing logic with init using Gaussian dist of original embeddings would fail when using DeepSpeed ZeRO3 as the weights are sharded and you would need to use `deepspeed.zero.GatheredParameters` for gathering the embedding layer weights before modifying them.

File: examples/pytorch/language-modeling/run_fim.py
Comment (tanaymeh) [reply examples/pytorch/language-modeling/run_fim.py:558]: Thanks for the review @pacman100!

I had a doubt, shouldn't I be using `deepspeed.zero.GatheredParameters` for that? It will look something like this:

```python
if is_deepspeed_zero3_enabled():
    import deepspeed
    with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=None):
            new_embeddings = embeddings.weight.clone()

            # Set the new tokens' embeddings to the newly sampled embeddings
            new_embeddings[-len(special_tokens) :] = new_token_embeddings
        
            # Update the model's embeddings with the new embeddings
            embeddings.weight = torch.nn.Parameter(new_embeddings)
            model.set_input_embeddings(embeddings)
```

**Comment by ArthurZucker** [comment]
@tanaymeh just waiting to make sure the deepspeed thing works and we should be able to merge

File: examples/pytorch/language-modeling/run_fim.py
Comment (pacman100) [reply examples/pytorch/language-modeling/run_fim.py:558]: Yes, I meant this, updated the original comment.

**Comment by tanaymeh** [comment]
@ArthurZucker @pacman100 Thanks for the reviews and sorry for the delay, I was busy with my coursework.

I have added the suggested changes and the deepspeed compatibility to both trainer and non-trainer scripts. Could you please review and merge?

TIA!

File: examples/pytorch/language-modeling/run_fim.py
Comment (pacman100) [inline examples/pytorch/language-modeling/run_fim.py:531]: Even `roiginal_embeddings` needs to be gathered when using DS ZeRO3 with zero init enabled.

File: examples/pytorch/language-modeling/run_fim.py
Comment (pacman100) [inline examples/pytorch/language-modeling/run_fim.py:561]:
```suggestion
        with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=0):
            # Set the new tokens' embeddings to the newly sampled embeddings
            embeddings.weight.data[-len(special_tokens) :] = new_token_embeddings
```

Comment (pacman100) [reply examples/pytorch/language-modeling/run_fim.py:561]: as the embeddings are being modified, the modifier rank must be given. Also, I think it is enough to just change the data.

**Review by pacman100 — COMMENTED** [review]
Hello @tanaymeh, wrt DeepSpeed few changes are still required. Please see the comment

File: examples/pytorch/language-modeling/run_fim.py
Comment (tanaymeh) [reply examples/pytorch/language-modeling/run_fim.py:561]: Thanks for the feedback @pacman100 

Is this below a good approach? I believe I will have to call GatheredParameters twice, once for original_embeddings and once for embeddings.

```python
    original_embeddings = model.get_input_embeddings()

    if is_deepspeed_zero3_enabled():
        import deepspeed
        with deepspeed.zero.GatheredParameters(original_embeddings.weight, modifier_rank=0):
            # Get the pre-expansion embeddings of the model and resize the embedding layer
            model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=pad_factor)
            embeddings = model.get_input_embeddings()

            # Sample the embeddings for the new tokens from a multivariate normal distribution
            # We do this so that the new embeddings are close to the original embeddings and not necessarily zero
            # More on this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html
            mean = original_embeddings.mean(dim=0)
            n = original_embeddings.size()[0]
            sigma = ((original_embeddings - mean).T @ (original_embeddings - mean)) / n
            dist = torch.distributions.multivariate_normal.MultivariateNormal(
                mean,
                covariance_matrix=1e-5 * sigma,
            )
            new_token_embeddings = torch.stack(
                tuple((dist.sample() for _ in range(len(special_tokens)))),
                dim=0,
            )
    else:
        original_embeddings = model.get_input_embeddings()
        # Get the pre-expansion embeddings of the model and resize the embedding layer
        model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=pad_factor)
        embeddings = model.get_input_embeddings()

        # Sample the embeddings for the new tokens from a multivariate normal distribution
        # We do this so that the new embeddings are close to the original embeddings and not necessarily zero
        # More on this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html
        mean = original_embeddings.mean(dim=0)
        n = original_embeddings.size()[0]
        sigma = ((original_embeddings - mean).T @ (original_embeddings - mean)) / n
        dist = torch.distributions.multivariate_normal.MultivariateNormal(
            mean,
            covariance_matrix=1e-5 * sigma,
        )
        new_token_embeddings = torch.stack(
            tuple((dist.sample() for _ in range(len(special_tokens)))),
            dim=0,
        )

    if is_deepspeed_zero3_enabled():
        import deepspeed
        with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=0):
            # Set the new tokens' embeddings to the newly sampled embeddings
            embeddings.weight.data[-len(special_tokens) :] = new_token_embeddings
    else:
        # Set the new tokens' embeddings to the newly sampled embeddings
        embeddings.weight.data[-len(special_tokens) :] = new_token_embeddings
    
    # Update the model's embeddings with the new embeddings
    model.set_input_embeddings(embeddings)
```

**Comment by ArthurZucker** [comment]
I already approved, I am just waiting for @pacman100's approval here! 🤗

File: examples/pytorch/language-modeling/run_fim.py
Comment (pacman100) [reply examples/pytorch/language-modeling/run_fim.py:561]: Hello @tanaymeh, the above changes are correct, Thank you!

**Review by pacman100 — APPROVED** [review]
Approving the PR, good to merge once the changes in the comment https://github.com/huggingface/transformers/pull/27464/files#r1505771256 are committed. Thank you @tanaymeh for all the work on adding examples for finetuning/pretraining code infilling models! 🔥🚀✨

**Comment by tanaymeh** [comment]
Thanks for approving the PR @ArthurZucker and @pacman100!

I have pushed all the changes and the PR should be ready to be merged 🤗

**Comment by amyeroberts** [comment]
@tanaymeh Thanks for working on this! I'll merge as I can see the requested changes are added and I'm on repo watch this week :)

**Pull request merged**
